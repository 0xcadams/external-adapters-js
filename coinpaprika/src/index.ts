import { expose, util } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'

const NAME = 'COINPAPRIKA'

export = { NAME, makeExecute, ...expose(util.wrapExecute(makeExecute())) }
