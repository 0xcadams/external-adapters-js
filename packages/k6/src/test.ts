import { check, sleep } from 'k6'
import { SharedArray } from "k6/data";
import http from 'k6/http'
import { Rate } from 'k6/metrics'
import {
  AdapterNames,
  ADAPTERS,
  GROUP_COUNT,
  httpPayloadsByAdapter,
  wsPayloads,
} from './config/index'

export const options = {
  vus: 1,
  duration: '12h',
  thresholds: {
    http_req_failed: ['rate<0.01'], // http errors should be less than 1%
    http_req_duration: ['p(95)<200'], // 95% of requests should be below 200ms
  },
}

let currIteration = 0
export const errorRate = new Rate('errors')

let payloadData = wsPayloads
if (__ENV.PAYLOAD_GENERATED) {
  payloadData = new SharedArray("payloadData", function () {
    // here you can open files, and then do additional processing or generate the array with data dynamically
    const f = JSON.parse(open("../src/config/ws.json"));
    return f; // f must be an array[]
  });
}

interface LoadTestGroupUrls {
  [loadTestGroup: string]: {
    [adapterName: string]: string
  }
}

function getLoadTestGroupsUrls(): LoadTestGroupUrls {
  if (__ENV.LOCAL_ADAPTER_NAME) {
    /**
     * Local environment only handles a single endpoint
     */
    return {
      local: {
        [__ENV.LOCAL_ADAPTER_NAME]: 'http://host.docker.internal:8080',
      },
    }
  } else {
    const loadTestGroup = Array(GROUP_COUNT)
      .fill(null)
      .map((_) => {
        if (__ENV.QA_RELEASE_TAG) {
          return `https://adapters.qa.stage.cldev.sh/`
        } else {
          return `https://adapters.main.stage.cldev.sh/`
        }
      })

    const adaptersToMap = ADAPTERS.filter((a) => currIteration % a.secondsPerCall === 0).map(
      (a) => a.name,
    )
    const adaptersPerLoadTestGroup = loadTestGroup.map(
      (u, i) =>
        [
          i,
          Object.fromEntries(
            adaptersToMap.map((a) => {
              if (__ENV.QA_RELEASE_TAG) {
                return [a, `${u}qa-ea-${a}-${__ENV.QA_RELEASE_TAG}`] as const
              }
              return [a, `${u}${a}`] as const
            }),
          ),
        ] as const,
    )

    return Object.fromEntries(adaptersPerLoadTestGroup)
  }
}

function buildRequests() {
  const batchRequests: Parameters<typeof http.batch>[0] = {}
  const params = {
    headers: {
      'Content-Type': 'application/json',
    },
  }
  const urls = getLoadTestGroupsUrls()
  for (const [loadTestGroup, adaptersByAdapterName] of Object.entries(urls)) {
    for (const [adapterName, url] of Object.entries(adaptersByAdapterName)) {
      if (__ENV.WS_ENABLED) {
        for (const payload of payloadData) {
          if (adapterName === 'coinapi') {
            const body = JSON.parse(payload.data)
            body.data.endpoint = 'assets'
            batchRequests[`Group-${loadTestGroup}-${adapterName}-${payload.name}`] = {
              method: payload.method,
              url,
              body,
              params,
            }
          } else {
            batchRequests[`Group-${loadTestGroup}-${adapterName}-${payload.name}`] = {
              method: payload.method,
              url,
              body: payload.data,
              params,
            }
          }
        }
      }

      for (const payload of httpPayloadsByAdapter[adapterName as AdapterNames]) {
        batchRequests[`Group-${loadTestGroup}-${adapterName}-${payload.name}`] = {
          method: payload.method,
          url,
          body: payload.data,
          params,
        }
      }
    }
  }

  return batchRequests
}

const batchRequests = buildRequests()

export default (): void => {
  currIteration++
  const responses = http.batch(batchRequests)
  for (const [name, response] of Object.entries(responses)) {
    const result = check(response, {
      [`${name} returned 200`]: (r) => r.status == 200,
    })

    errorRate.add(!result)
  }

  sleep(1)
}
