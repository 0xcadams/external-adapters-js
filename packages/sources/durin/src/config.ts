import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig } from '@chainlink/types'

export const NAME = 'DURIN'

export const DEFAULT_ENDPOINT = 'gateway'
export const DEFAULT_BASE_URL = 'http://localhost:18081'

export interface Config extends BaseConfig {
  l2RpcUrl: string
  addressManagerContract: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix)
  config.defaultEndpoint = DEFAULT_ENDPOINT
  config.rpcUrl = util.getRequiredEnv('RPC_URL')
  return {
    ...Requester.getDefaultConfig(prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
    rpcUrl: util.getRequiredEnv('RPC_URL', prefix),
    l2RpcUrl: util.getRequiredEnv('L2_RPC_URL', prefix),
    addressManagerContract: util.getRequiredEnv('ADDRESS_MANAGER_CONTRACT', prefix),
  }
}
