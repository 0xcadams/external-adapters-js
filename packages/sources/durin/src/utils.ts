import { Fragment, Interface, JsonFragment } from '@ethersproject/abi'

export const toInterface = (
  abi: string | readonly (string | Fragment | JsonFragment)[] | Interface,
) => {
  if (Interface.isInterface(abi)) {
    return abi
  }
  return new Interface(abi)
}
