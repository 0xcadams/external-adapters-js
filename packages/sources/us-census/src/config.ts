import { Requester } from '@chainlink/ea-bootstrap'
import { Config as ChainlinkConfig } from '@chainlink/types'

export const NAME = 'US_CENSUS'

export const DEFAULT_ENDPOINT = 'census'

export const makeConfig = (prefix?: string): ChainlinkConfig => {
  const defaultConfig = Requester.getDefaultConfig(prefix, true)

  const config: ChainlinkConfig = {
    ...defaultConfig,
    defaultEndpoint: DEFAULT_ENDPOINT,
  }

  return config
}
