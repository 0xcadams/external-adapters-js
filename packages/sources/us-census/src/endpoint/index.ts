export * as census from './census'
