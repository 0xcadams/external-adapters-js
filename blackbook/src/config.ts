import { Requester, logger } from '@chainlink/external-adapter'
import { util } from '@chainlink/ea-bootstrap'
import types from '@chainlink/types'
import { NAME } from './endpoint/vehicle'

const ENV_API_USERNAME = 'API_USERNAME'
const ENV_API_PASSWORD = 'API_PASSWORD'

export const DEFAULT_ENDPOINT = NAME

export type Config = types.Config & {
  username: string
  password: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix)
  config.api.baseURL = config.api.baseURL || 'https://service.blackbookcloud.com/'
  return {
    ...config,
    username: util.getRequiredEnv(ENV_API_USERNAME),
    password: util.getRequiredEnv(ENV_API_PASSWORD),
  }
}

// Config without sensitive data
const redact = (config: Config) => ({
  ...config,
  apiKey: '*****',
  username: '*****',
  password: '*****',
})

export function logConfig(config: Config): void {
  logger.debug('Adapter configuration:', { config: config && redact(config) })
}
