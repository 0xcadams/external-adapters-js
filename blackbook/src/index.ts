import { expose, util } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'blackbook'

export = { NAME, makeExecute, makeConfig, ...expose(util.wrapExecute(makeExecute())) }
