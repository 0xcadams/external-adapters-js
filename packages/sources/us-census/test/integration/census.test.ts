import { AdapterRequest } from '@chainlink/types'
import request from 'supertest'
import process from 'process'
import nock from 'nock'
import http from 'http'
import { server as startServer } from '../../src'
import { mockResponseWithInvalidLatitude, mockResponseWithInvalidVariable } from './fixtures'

let oldEnv: NodeJS.ProcessEnv

process.env.API_KEY = 'test_api_key'

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.API_VERBOSE = process.env.API_VERBOSE || 'true'
  if (process.env.RECORD) {
    nock.recorder.rec()
  }
})

afterAll(() => {
  process.env = oldEnv
  if (process.env.RECORD) {
    nock.recorder.play()
  }

  nock.restore()
  nock.cleanAll()
  nock.enableNetConnect()
})

describe('execute', () => {
  const id = '1'
  let server: http.Server
  const req = request('localhost:8080')
  beforeAll(async () => {
    server = await startServer()
  })
  afterAll((done) => {
    server.close(done)
  })

  describe('with invalid geography', () => {
    const data: AdapterRequest = {
      id,
      data: {
        endpoint: 'acs5_2019',
        geography: 'tract',
        variables: ['B25001_001E', 'B25002_002E'],
        latitude: -1,
        longitude: -122.419418,
      },
    }

    mockResponseWithInvalidLatitude()

    it('should return error', async () => {
      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(400)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('with invalid variable', () => {
    const data: AdapterRequest = {
      id,
      data: {
        endpoint: 'acs5_2019',
        geography: 'state',
        variables: ['SOME_INVALID_VAR'],
        latitude: 37.774929,
        longitude: -122.419418,
      },
    }

    mockResponseWithInvalidVariable()

    it('should return error', async () => {
      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(400)
      expect(response.body).toMatchSnapshot()
    })
  })
})
