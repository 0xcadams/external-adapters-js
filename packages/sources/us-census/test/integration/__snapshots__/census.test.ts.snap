// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute with invalid geography should return error 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"acs5_2019\\",\\"geography\\":\\"tract\\",\\"variables\\":[\\"B25001_001E\\",\\"B25002_002E\\"],\\"latitude\\":-1,\\"longitude\\":-122.419418}}",
    "message": Object {
      "error": Object {
        "message": "No FIPS (Census geocodes) found for given arguments",
        "name": "AdapterError",
      },
      "jobRunID": "1",
      "status": "errored",
      "statusCode": 400,
    },
    "name": "AdapterError",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`execute with invalid variable should return error 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"endpoint\\":\\"acs5_2019\\",\\"geography\\":\\"state\\",\\"variables\\":[\\"SOME_INVALID_VAR\\"],\\"latitude\\":37.774929,\\"longitude\\":-122.419418}}",
    "message": Object {
      "error": Object {
        "message": "STATUS: 400  for: https://api.census.gov/data/2019/acs/acs5?get=NAME,SOME_INVALID_VAR&for=state:06",
        "name": "AdapterError",
      },
      "jobRunID": "1",
      "status": "errored",
      "statusCode": 400,
    },
    "name": "AdapterError",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 400,
}
`;
